### PR TITLE
Add recipe for Leptonica

### PR DESCRIPTION
**Pull request by CurtLH**

### CODE REVIEW

**Comment by CurtLH** [comment]
I have `openjpeg` listed in both `host` and `run`, but it seems like it cannot find it. 

```bash
-- Looking for include file openjpeg-2.0/openjpeg.h
-- Looking for include file openjpeg-2.0/openjpeg.h - not found
-- Looking for include file openjpeg-2.1/openjpeg.h
-- Looking for include file openjpeg-2.1/openjpeg.h - not found
-- Looking for include file openjpeg-2.2/openjpeg.h
-- Looking for include file openjpeg-2.2/openjpeg.h - not found
-- Looking for include file openjpeg-2.3/openjpeg.h
-- Looking for include file openjpeg-2.3/openjpeg.h - not found
```

File: recipes/leptonica/build.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+mkdir build && cd build
+cmake -DCMAKE_INSTALL_PREFIX:PATH=$PREFIX ..
```
Comment (isuruf) [inline recipes/leptonica/build.sh]: Add `-DCMAKE_PREFIX_PATH=$PREFIX -DCMAKE_BUILD_TYPE=Release` here

**Comment by CurtLH** [comment]
Thanks for fixing my mistake @isuruf!

**Comment by msarahan** [comment]
This is annoying.  It seems like there might be a bug in the CMake scripts that look for openjpeg:

```
-- Checking for module 'libopenjp2'
--   Found libopenjp2, version 2.3.0
```

but it's clearly not setting an include path. Probably because it's looking in these subfolders:

```
-- Looking for include file openjpeg-2.0/openjpeg.h
-- Looking for include file openjpeg-2.0/openjpeg.h - not found
-- Looking for include file openjpeg-2.1/openjpeg.h
-- Looking for include file openjpeg-2.1/openjpeg.h - not found
-- Looking for include file openjpeg-2.2/openjpeg.h
-- Looking for include file openjpeg-2.2/openjpeg.h - not found
-- Looking for include file openjpeg-2.3/openjpeg.h
-- Looking for include file openjpeg-2.3/openjpeg.h - not found
``` 

I see the openjpeg-2.3 folder with that file in the package, but I don't know what base include path cmake is looking in.  You may need to set CMAKE_INCLUDE_PATH, but we never have to do that in Anaconda recipes.  Could be an issue with the toolchain package's include addition.  Conda-forge's toolchain does it in CPPFLAGS: https://github.com/conda-forge/toolchain-feedstock/blob/master/recipe/activate.sh#L43

For Anaconda, we have that in CFLAGS: https://github.com/AnacondaRecipes/aggregate/blob/master/ctng-compilers-activation-feedstock/recipe/activate-gcc.sh#L87

I don't know if this is actually broken - I've seen include stuff in both places all over the net - but it is the only real difference I see.  You might need to get into CMake's verbose output, https://stackoverflow.com/questions/2670121/using-cmake-with-gnu-make-how-can-i-see-the-exact-commands to learn more about exactly what include paths are being used.

**Comment by mingwandroid** [comment]
It's likely that `conda-forge` need my changes to `pkg-config` for this to work correctly (iff openjpg has `.pc` files that is); may also need my changes to `libtool`.

**Comment by CurtLH** [comment]
@bgruening -- I noticed that you are a contributor on the [bioconda recipe for leptonica](https://github.com/bioconda/bioconda-recipes/blob/1f9ec5890d7d67c075a134906796ffc0962ceeae/recipes/leptonica/1.73/meta.yaml).  Do you have any thoughts regarding the issues mentioned in this PR (and the [PR for tesseract](https://github.com/conda-forge/staged-recipes/pull/5935)).

**Comment by sodre** [comment]
@mingwandroid, 
  What was the fix for pkg-config that you did? I think that is the cause of the problem, because I was able to get leptonica to compile using anaconda/pkg-config instead of the one in conda-forge.

Side-question: If there is a way in meta.yaml to force pkg-config to be picked up from anaconda instead of conda-forge, I can push the recipe so we can all verify that it works.

**Comment by isuruf** [comment]
There were 2 ways implemented to find openjpeg in the CMake files.
1. Use pkg-config, but they used `<PREFIX>_INCLUDE_DIR` instead of `<PREFIX>_INCLUDE_DIRS`.
2. Check in compiler's implicit include directories for `openjpeg-2.0/openjpeg.h`, `openjpeg-2.1/openjpeg.h`, etc. (Anaconda 5 compilers have $PREFIX and therefore it succeeds, but toolchain doesn't.

Fix for 1 is,

```diff
commit c74e712799890a600814d4d4970f8f03be254b0a
Author: Isuru Fernando <isuruf@gmail.com>
Date:   Wed May 30 19:29:02 2018 -0600

    Fix pkgconfig variables

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 96ca496..945100a 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -40,7 +40,7 @@ if (JPEG_LIBRARY)
     target_link_libraries       (leptonica ${JPEG_LIBRARY})
 endif()
 if (JP2K_FOUND)
-    target_include_directories  (leptonica PUBLIC ${JP2K_INCLUDE_DIR})
+    target_include_directories  (leptonica PUBLIC ${JP2K_INCLUDE_DIRS})
     target_link_libraries       (leptonica ${JP2K_LIBRARIES})
 endif()
 if (PNG_LIBRARY)
@@ -52,7 +52,7 @@ if (TIFF_LIBRARY)
     target_link_libraries       (leptonica ${TIFF_LIBRARY})
 endif()
 if (WEBP_FOUND)
-    target_include_directories  (leptonica PUBLIC ${WEBP_INCLUDE_DIR})
+    target_include_directories  (leptonica PUBLIC ${WEBP_INCLUDE_DIRS})
     target_link_libraries       (leptonica ${WEBP_LIBRARIES})
 endif()
 if (ZLIB_LIBRARY)
```

**Comment by sodre** [comment]
@isuruf nice catch!

Should we open upstream bug-report, or this something related to our distro?

**Comment by isuruf** [comment]
It's definitely a bug in their cmake files. If somebody can send a PR upstream with that patch that'd be great.

File: recipes/leptonica/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    - {{ compiler('c') }}
+  host:
+    - cmake
+    - pkg-config
+    - libwebp 
+    - zlib
+    - libpng
+    - libtiff
+    - jpeg
+    - openjpeg
```
Comment (isuruf) [inline recipes/leptonica/meta.yaml]: Add `giflib` here and in run too.

File: recipes/leptonica/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+build:
+  number: 0
+  skip: True  # [osx or win]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+  host:
+    - cmake
+    - pkg-config
```
Comment (isuruf) [inline recipes/leptonica/meta.yaml]: Move these 2 packages to build

File: recipes/leptonica/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+{% set sha256 = "ff98ffef6103cda0e400676f81113a8157569ac0f453c6e03a1a27b305ba25a7" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/DanBloomberg/{{ name }}/archive/{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+  patches: c74e712799890a600814d4d4970f8f03be254b0a.patch # fixes openjpeg
```
Comment (isuruf) [inline recipes/leptonica/meta.yaml]: Can you send a PR upstream and add a comment with the link?

File: recipes/leptonica/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+build:
+  number: 0
+  skip: True  # [osx or win]
```
Comment (isuruf) [inline recipes/leptonica/meta.yaml]: Do you want to try enabling osx?

File: recipes/leptonica/meta.yaml (hunk 2 above)
Comment (CurtLH) [reply recipes/leptonica/meta.yaml]: Done

File: recipes/leptonica/meta.yaml (hunk 4 above)
Comment (CurtLH) [reply recipes/leptonica/meta.yaml]: Will do.

File: recipes/leptonica/meta.yaml (hunk 5 above)
Comment (CurtLH) [reply recipes/leptonica/meta.yaml]: Done

**Comment by isuruf** [comment]
Looks like their cmake files are not complete. Can you try switching to an autotools build? It seems the autotools build is much more mature than the cmake build.

**Comment by CurtLH** [comment]
@isuruf -- I'm not very familiar with the process of building these types of packages.  Can you point me in the direction of an example using autotools build?  Or maybe even just some keywords to search for that might lead me in the right direction.

File: recipes/leptonica/meta.yaml (hunk 4 above)
Comment (CurtLH) [reply recipes/leptonica/meta.yaml]: https://github.com/DanBloomberg/leptonica/pull/345

File: recipes/leptonica/build.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+./autobuild
+./configure --prefix=$PREFIX
+make -j$CPU_COUNT
+make install
+#mkdir build && cd build
+#cmake -DCMAKE_INSTALL_PREFIX:PATH=$PREFIX -DCMAKE_PREFIX_PATH=$PREFIX -DCMAKE_BUILD_TYPE=Release ..
+#make install
```
Comment (sodre) [inline recipes/leptonica/build.sh]: Remove the commented lines.

**Comment by mingwandroid** [comment]
Ongoing @mbargull

**Comment by mbargull** [comment]
Feel free to merge. This is clearly a non-bio package and thus belongs here. We'll remove its recipe from Bioconda.

**Comment by CurtLH** [comment]
I think this one is ready to merge @conda-forge/staged-recipes!

**Pull request merged**